File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        n=0
        # cal mid node by slow and fast
        slow=head
        fast=head
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next


        second=slow.next
        slow.next=None
        # reverse second  linked list
        prev=None
        while second:
            nxt=second.next
            second.next=prev
            prev=second
            second=nxt

        first = head
        second = prev
        while second:
            nxt1=first.next
            nxt2=second.next

            first.next=second
            second.next=nxt1

            first = nxt1
            second=nxt2

        # TC O(n) AND SC O(1)
            



        

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head){
            return NULL;
        }

        int l=0;
        ListNode * temp=head;
        while(temp){
            temp=temp->next;
            l++;
        }

        if (n==l){
            return head->next;
        }

        ListNode* prev=NULL;
        temp=head;
        int idx = l-n;
        while(idx--){
            prev=temp;
            temp=temp->next;
        }
        ListNode * nodeToDelete=temp;
        prev->next=temp->next;
        delete nodeToDelete;
        // TC O(N) AND SCO(1)
        return head;
    }
};
